app/src/main/java/protect/card_locker/LetterBitmap.kt

```diff
@@ -0,0 +1,133 @@
+) {
+    /**
+     * A {@link Bitmap} that contains a letter used in the English
+     * alphabet or digit, if there is no letter or digit available, a
+     * default image is shown instead
+     */
+    private val letterTile: Bitmap
+
+    /**
+     * The background color of the letter bitmap
+     */
+    private val mColor: Int
+
+    init {
+        val paint = TextPaint().apply {
+            color = textColor ?: Color.WHITE
+            textAlign = Paint.Align.CENTER
+            isAntiAlias = true
+            textSize = tileLetterFontSize.toFloat()
+            typeface = Typeface.defaultFromStyle(Typeface.BOLD)
+        }
+
+        mColor = backgroundColor ?: getDefaultColor(context, key)
+
+        this.letterTile = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
+
+        var firstChar = displayName.substring(0, 1).uppercase(Locale.getDefault())
+        var firstCharEnd = 2
+        while (firstCharEnd <= displayName.length) {
+            // Test for the longest render-able string
+            // But ignore containing only a-Z0-9 to not render things like ffi as a single character
+            val test = displayName.substring(0, firstCharEnd)
+            if (!isAlphabetical(test) && PaintCompat.hasGlyph(paint, test)) {
+                firstChar = test
+            }
+            firstCharEnd++
+        }
+
+        Log.d(
+            "LetterBitmap",
+            "using sequence $firstChar to render first char which has length ${firstChar.length}"
+        )
+
+        Canvas().apply {
+            setBitmap(this@LetterBitmap.letterTile)
+            drawColor(mColor)
+
+            val bounds = Rect()
+            paint.getTextBounds(firstChar, 0, firstChar.length, bounds)
+            drawText(
+                firstChar,
+                0, firstChar.length,
+                width / 2.0f, (height - (bounds.bottom + bounds.top)) / 2.0f,
+                paint
+            )
+        }
+    }
+
+    val backgroundColor: Int
+        /**
+         * @return background color used for letter title.
+         */
+        get() = mColor
+
+    companion object {
+        /**
+         * @param key The key used to generate the tile color
+         * @return A new or previously chosen color for `key` used as the
+         * tile background color
+         */
+        private fun pickColor(key: String, colors: TypedArray): Int {
+            // String.hashCode() is not supposed to change across java versions, so
+            // this should guarantee the same key always maps to the same color
+            val color = abs(key.hashCode()) % colors.length()
+            return colors.getColor(color, Color.BLACK)
+        }
+
+        private fun isAlphabetical(string: String): Boolean {
+            return string.matches("[a-zA-Z0-9]*".toRegex())
+        }
+
+        /**
+         * Determine the color which the letter tile will use if no default
+         * color is provided.
+         */
+        @JvmStatic
```

Comment:
The force pushing (while I'm fine with it) seems to be confusing the GitHub UI. Yes, please use .Companion() in Utils instead of an annotation like this, because annotations are easy to forget during future refactors :)